src/dotnet/commands/dotnet-build/CompileContext.cs

```diff
@@ -514,31 +515,40 @@ private void MakeRunnable()
             executable.MakeCompilationOutputRunnable();
         }
 
-        private static ISet<ProjectDescription> Sort(Dictionary<string, ProjectDescription> projects)
+        private static IEnumerable<ProjectDescription> Sort(ProjectDependenciesFacade dependencies)
         {
-            var outputs = new HashSet<ProjectDescription>();
+            var outputs = new List<ProjectDescription>();
 
-            foreach (var pair in projects)
+            foreach (var pair in dependencies.Dependencies)
             {
-                Sort(pair.Value, projects, outputs);
+                Sort(pair.Value, dependencies, outputs);
             }
 
-            return outputs;
+            return outputs.Distinct(new ProjectComparer());
         }
 
-        private static void Sort(ProjectDescription project, Dictionary<string, ProjectDescription> projects, ISet<ProjectDescription> outputs)
+        private static void Sort(LibraryExport node, ProjectDependenciesFacade dependencies, IList<ProjectDescription> outputs)
         {
             // Sorts projects in dependency order so that we only build them once per chain
-            foreach (var dependency in project.Dependencies)
+            ProjectDescription projectDependency;
+            foreach (var dependency in node.Library.Dependencies)
             {
-                ProjectDescription projectDependency;
-                if (projects.TryGetValue(dependency.Name, out projectDependency))
-                {
-                    Sort(projectDependency, projects, outputs);
-                }
+                // Sort the children
+                Sort(dependencies.Dependencies[dependency.Name], dependencies, outputs);
             }
 
-            outputs.Add(project);
+            // Add this node to the list if it is a project
+            if (dependencies.ProjectDependenciesWithSources.TryGetValue(node.Library.Identity.Name, out projectDependency))
+            {
+                // Add to the list of projects to build
+                outputs.Add(projectDependency);
+            }
+        }
+
+        private class ProjectComparer : IEqualityComparer<ProjectDescription>
+        {
+            public bool Equals(ProjectDescription x, ProjectDescription y) => string.Equals(x.Identity.Name, y.Identity.Name);
```

Comment:
Oh I take that back... string.Equals does do ordinal.  I was thinking of StartsWith, IndexOf, etc.  My mistake.  Carry on, nothing to see here.